rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,737 @@
+
+An example of the file layout for the resource used as an example in this RFD is included below.
+
+```bash
+api/proto
+├── README.md
+├── buf-legacy.yaml
+├── buf.lock
+├── buf.yaml
+└── teleport
+    ├── foo
+    │   └── v1
+    │       ├── foo.proto
+    │       └── foo_service.proto
+    ├── legacy
+    │   ├── client
+    │   │   └── proto
+    │   │     ├── authservice.proto
+    │   │     ├── certs.proto
+    │   │     ├── event.proto
+    │   │     ├── joinservice.proto
+    │   │     └── proxyservice.proto
+    │   └── types
+    │         ├── device.proto
+    │         ├── events
+    │         │    ├── athena.proto
+    │         │    └── events.proto
+    │         ├── types.proto
+    │         ├── webauthn
+    │         │    └── webauthn.proto
+    │         └── wrappers
+    │              └── wrappers.proto
+```
+
+The legacy directory contains resources and API definitions that were defined prior to our shift to user smaller,
+localized services per resource. Adding new resources and APIs to the giant monolithic `proto.AuthService` should be
+avoided if possible.
+
+### New resource checklist
+
+Use the non-exhaustive list below as a guide when adding a new resource. Further sections in the RFD dive into more
+detail about what is needed to complete a particular item.
+
+- [ ] Create proto spec for resource and RPC service
+- [ ] Create backend service
+- [ ] Add resource client to `api` client
+- [ ] Implement gRPC service
+- [ ] Add resource support to tctl (get, create, edit)
+- [ ] Optional: Add resource to cache
+- [ ] Optional: Add support for bootstrapping the resource
+- [ ] Optional: Add support for resource to Teleport Operator
+- [ ] Optional: Add support for resource to Teleport Terraform Provider
+
+### Defining a resource
+
+A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
+additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
+derive from the message definition itself, they need to be defined so that anything processing a generic resource can
+identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
+and leverages `services.UnknownResource` to identify the resource and act appropriately.
+
+```protobuf
+import "teleport/header/v1/resourceheader.proto";
+
+// Foo is a resource that does foo.
+message Foo {
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The specific properties of a Foo.
+  FooSpec spec = 5;
```

Comment:
From a Machine ID perspective, I'd certainly much prefer a typed `status` field to store things like the cert revision in rather than the current solution using the metadata.

Generally, I like the idea of a `status` field, especially in terms of how we interact with IaC and it allows us to clearly mark which fields should not be modified by users.

I suppose an alternative option is to just store these sorts of things in another resource - but this adds a lot of complexity.

Reply:
Are we OK having `spec` fields that are solely managed by Teleport, or should we consider these a part of `status` too?

Reply:
In my opinion, we have two cases:
- the resource is created by the user/exposed to the user -> everything in spec should be managed by the user, if something is managed by teleport, it should be out of spec
- the resource is not created by the user/exposed via `tctl` -> we do what we want

If we have fields in the spec that are managed by teleport and some managed by the user, it means we'll have to manually tell each IaC tool to skip reconciling those fields (and we'll forget some fields and end up with something broken).

Reply:
I think the Database resource is a good example of this:
There's a `spec` and a `status`:
https://github.com/gravitational/teleport/blob/464edfa2cdc7acc4e8d2b711fef3e9740e21d4d5/api/proto/teleport/legacy/types/types.proto#L260-L268

Reply:
LGTM!